src/python/pants/backend/jvm/subsystems/zinc.py

```diff
@@ -73,3 +108,81 @@ def sbt_jar(name, **kwargs):
                           main='no.such.main.Main',
                           custom_rules=shader_rules,
                           **kwargs)
+
+    cls.register_jvm_tool(register,
+                          Zinc.ZINC_EXTRACTOR_TOOL_NAME,
+                          classpath=[
+                            JarDependency('org.pantsbuild', 'zinc-extractor_2.11', '0.0.1')
+                          ])
+
+  def __init__(self, *args, **kwargs):
+    super(Zinc, self).__init__(*args, **kwargs)
+    self.set_distribution(jdk=True)
+
+  @memoized_property
+  def rebase_map_args(self):
+    """We rebase known stable paths in zinc analysis to make it portable across machines."""
+    rebases = {
+        self.dist.real_home: '/dev/null/remapped_by_pants/java_home/',
+        get_buildroot(): '/dev/null/remapped_by_pants/buildroot/',
+        self.get_options().pants_workdir: '/dev/null/remapped_by_pants/workdir/',
+      }
+    return (
+        '-rebase-map',
+        ','.join('{}:{}'.format(src, dst) for src, dst in rebases.items())
+      )
+
+  @classmethod
+  def _compiler_plugins_cp_entries(cls, jvm_tool_mixin_instance, products):
+    """Any additional global compiletime classpath entries.
+
+    TODO: Switch to instance memoized_property after 1.6.0.dev0.
+    """
+    def cp(toolname):
+      scope = jvm_tool_mixin_instance.options_scope
+      return jvm_tool_mixin_instance.tool_classpath_from_products(products,
+                                                                  toolname,
+                                                                  scope=scope)
+    classpaths = cp('javac-plugin-dep') + cp('scalac-plugin-dep')
+    return [(conf, jar) for conf in cls.DEFAULT_CONFS for jar in classpaths]
+
+  def extractor_classpath(self, products):
+    return self.tool_classpath_from_products(products,
+                                             self.ZINC_EXTRACTOR_TOOL_NAME,
+                                             scope=self.options_scope)
+
+  def compile_classpath(self, products, classpath_product_key, target, extra_cp_entries=None):
+    """Compute the compile classpath for the given target."""
+    return Zinc.compile_classpath_for(self,
+                                      products,
+                                      classpath_product_key,
+                                      target,
+                                      extra_cp_entries=extra_cp_entries)
+
+  @classmethod
+  def compile_classpath_for(cls,
```

Comment:
I had to read this a couple times to figure out how the arguments passed by the caller map to these params. It feels unusual to have a classmethod take an instance of the class as an argument. It is not using it as a bootstrap in a factory so I am a bit in the dark as to what is motivating that. Just seems like it would be a better fit as an instance method or something. 

I am just scrolling through in a web browser so forgive dumb questions. I also see some "merge with X when 1.6.0", so I assume there is some sort of bug/deprecation that is forcing your hand.